app/src/gecko/java/org/mozilla/focus/web/WebViewProvider.java

```diff
@@ -467,12 +471,36 @@ public boolean canGoBack() {
 
         @Override
         public void restoreWebViewState(Session session) {
-            // TODO: restore navigation history, and reopen previously opened page
+            final Bundle stateData = session.getWebViewState();
+            final String desiredURL = session.getUrl().getValue();
+            final GeckoSession.SessionState sessionState = stateData.getParcelable("state");
+            if (sessionState != null) {
+                geckoSession.restoreState(sessionState);
+            } else {
+                loadUrl(desiredURL);
+            }
         }
 
         @Override
-        public void saveWebViewState(@NonNull Session session) {
-            // TODO: save anything needed for navigation history restoration.
+        public void saveWebViewState(@NonNull final Session session, final CountDownLatch latch) {
```

Comment:
Let's not expose the CountDownLatch in the API on `IWebView`. You should be able to create the latch here, post the runnable and then wait .. all inside saveWebViewState here?